text/0496-handlebars-strict-mode.md

```diff
@@ -0,0 +1,591 @@
+
+[RFC #432](https://github.com/emberjs/rfcs/blob/master/text/0432-contextual-helpers.md#relationship-with-globals)
+described some additional issues with the current implicit globals semantics.
+
+We propose to remove support for implicit globals in strict mode. All values
+must be explicitly brought into scope, either through block params or defined
+in the "ambient scope" (see [Detailed design](#detailed-design) section).
+
+### 2. No implicit `this` fallback
+
+Today, when Ember sees a path like `{{foo}}` in the template, after exhausting
+the possibilities of implicit globals, it falls back to `{{this.foo}}`. This
+adds to the same confusion outlined above. More details about the motivation
+can be found in the accepted [RFC #308](https://github.com/emberjs/rfcs/pull/308).
+
+We propose to remove support for implicit `this` fallback in strict mode. The
+explicit form `{{this.foo}}` must be used to refer to instance state, otherwise
+it will trigger the same errors mentioned in the previous section.
+
+It is worth mentioning that [RFC #432](https://github.com/emberjs/rfcs/pull/432)
+laid out a transition path towards a world where "everything is a value".
+Between this and the previous restriction, we essentially have completed that
+transition. To recap, here is a list of the
+[outstanding issues](https://github.com/emberjs/rfcs/blob/master/text/0432-contextual-helpers.md#relationship-with-globals):
+
+1. Not possible to reference globals outside of invocation positions
+2. Invocation of global helpers in angle bracket named arguments positions
+3. Naming collisions between global components, helpers and element modifiers
+
+All of these problems are all related to implicit globals and/or implicit
+`this` fallback. Since neither of these features are supported in strict mode,
+they are no longer a concern for us.
+
+### 3. No implicit invocation of argument-less helpers
+
+In the contextual helpers RFC, we discussed [an issue](https://github.com/emberjs/rfcs/blob/master/text/0432-contextual-helpers.md#relationship-with-globals)
+regarding the invocation of arguments-less helpers in invocation argument
+positions.
+
+In today's semantics, if there is a global helper named `pi`, the following
+template will result it the helper being invoked, and the result of the helper
+invocation will be passed into the component.
+
+```hbs
+<MyComponent @value={{pi}} />
+```
+
+This is not desirable behavior in the value-based semantics proposed in that
+RFC, because it makes it impossible to pass helpers around as values, just as
+it is possible to pass around contextual components today.
+
+The contextaul helper RFC proposed to deprecate this behavior and require
+mandatory parentheses to invoke the helper.
+
+
+```hbs
+{{!-- passing the helper pi to the component --}}
+<MyComponent @value={{pi}} />
+
+{{!-- passing the result of invoking the helper pi to the component --}}
+<MyComponent @value={{(pi)}} />
```

Comment:
Just to be clear, this syntax is not supported yet, right? For me, it throws a parse error, e.g.:

```hbs
<MyComponent @date={{(date-now)}} />
```

Reply:
Yes! We will have to make it work before we can deprecate